homeassistant/components/homematic.py

```diff
@@ -0,0 +1,528 @@
+                             remoteport=remote_port,
+                             systemcallback=system_callback_handler,
+                             interface_id="homeassistant")
+
+    # Start server thread, connect to peer, initialize to receive events
+    HOMEMATIC.start()
+
+    # Stops server when Homeassistant is shutting down
+    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, HOMEMATIC.stop)
+    hass.config.components.append(DOMAIN)
+
+    return True
+
+
+# pylint: disable=too-many-branches
+def system_callback_handler(src, *args):
+    """Callback handler."""
+    if src == 'newDevices':
+        # pylint: disable=unused-variable
+        (interface_id, dev_descriptions) = args
+        key_dict = {}
+        # Get list of all keys of the devices (ignoring channels)
+        for dev in dev_descriptions:
+            key_dict[dev['ADDRESS'].split(':')[0]] = True
+        # Connect devices already created in HA to pyhomematic and
+        # add remaining devices to list
+        devices_not_created = []
+        for dev in key_dict:
+            try:
+                if dev in HOMEMATIC_DEVICES:
+                    for hm_element in HOMEMATIC_DEVICES[dev]:
+                        hm_element.link_homematic()
+                else:
+                    devices_not_created.append(dev)
+            # pylint: disable=broad-except
+            except Exception as err:
+                _LOGGER.error("Failed to setup device %s: %s", str(dev),
+                              str(err))
+        # If configuration allows autodetection of devices,
+        # all devices not configured are added.
+        if HOMEMATIC_AUTODETECT and devices_not_created:
+            for component_name, discovery_type in (
+                    ('switch', DISCOVER_SWITCHES),
+                    ('light', DISCOVER_LIGHTS),
+                    ('rollershutter', DISCOVER_ROLLERSHUTTER),
+                    ('binary_sensor', DISCOVER_BINARY_SENSORS),
+                    ('sensor', DISCOVER_SENSORS),
+                    ('thermostat', DISCOVER_THERMOSTATS)):
+                # Get all devices of a specific type
+                try:
+                    found_devices = _get_devices(discovery_type,
+                                                 devices_not_created)
+                # pylint: disable=broad-except
+                except Exception as err:
+                    _LOGGER.error("Failed generate opt %s with error '%s'",
+                                  component_name, str(err))
+
+                # When devices of this type are found
+                # they are setup in HA and an event is fired
+                if found_devices:
+                    try:
+                        component = get_component(component_name)
+                        config = {component.DOMAIN: found_devices}
+
+                        # Ensure component is loaded
+                        homeassistant.bootstrap.setup_component(
```

Comment:
Instead of line 165-180, you can use:

```python
from homeassistant.helpers import discovery

discovery.load_platform(hass, component, DOMAIN, {
    ATTR_DISCOVERED_DEVICES: found_devices
}, config)
```

Please note here that it is _very_ important that you pass the config that you get in `setup(hass, config)`. If you don't, you will break configs because for example the users switch config is now being ignored!